### PR TITLE
userdbd,homed,importd,sysupdate,pid1: make it easier to run daemons that have callouts from build dir

### PR DESCRIPTION
**Pull request by poettering**
We have a couple of daemons now which have a very intimate relationship to some callout binaries. It's a bit annoying to debug these with "gdb" directly form the build tree, since they will generally invoke the OS installed versions of their callout binaries, rather than those in the build tree. This has been getting on my nerves for a while.

Let's try to do something about this systematically, and automatically detect if we are run from a build tree or from an OS, and do the right thing in each case. This is bound DT_RPATH/DT_RUNPATH ELF logic: the assumption is that if these fields are set, we prefer build tree binaries over installed binaries, and if they aren't then not.

This takes benefit of the fact that meson sets rpath/runpath in the build tree and strips it during install. Hooking into this logic to find callout binaries makes a ton of sense, since if shared libraries are preferably found at some project-specific place, then binaries really should be treated similar.

TLDR: with this you can run homed/importd/userdbd/sysupdated/pid directly from the meson build tree and the right callouts will magically be used.

### CODE REVIEW

File: src/basic/build-path.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        r = get_build_exec_dir(&build_dir);
+        if (r < 0)
+                return r;
+
+        _cleanup_free_ char *fn = NULL;
+        r = path_extract_filename(path, &fn);
+        if (r < 0)
+                return r;
+        if (r == O_DIRECTORY) /* Uh? */
+                return -EISDIR;
```
Comment (YHNdnzj) [inline src/basic/build-path.c]: Hmm, this is duplicated in `find_environment_binary` too. Maybe move it to `invoke_callout_binary`?
Also, I think we should refuse `O_DIRECTORY` with `assert`. The original path is provided by us after all, and thus it is a programming error.

Comment (poettering) [reply src/basic/build-path.c]: i moved this now, but I left the EISDIR in it. While right now we just pass a hardcoded path in each case I think there's value in keeping this somewhat generic, and in allow possibly user-controlled paths for this too. Or to say this differently: if this was "main program" code I think an assert() would make a ton of sense, since really know how we are calling this. But given this is "library-like" code, I think we should keep a modicum of flexibility here, and rather refuse bad input politely than assert on it brutally. If that makes any sense?

**Review by bluca — APPROVED** [review]
